File: apply_migration.py
#!/usr/bin/env python3
"""
Apply Migration 021 to Supabase VPS Database
This script applies the complete monetization system migration.
"""

import psycopg2
import sys

# VPS Database Configuration
DB_CONFIG = {
    'host': '89.117.60.144',
    'port': 54322,
    'database': 'postgres',
    'user': 'postgres',
    'password': 'postgres'
}

MIGRATION_FILE = 'packages/supabase/supabase/migrations/021_monetization_system.sql'

def main():
    print('=' * 60)
    print('Applying Migration 021: Monetization System')
    print('=' * 60)
    print()

    # Read migration file
    print('[1/3] Reading migration file...')
    try:
        with open(MIGRATION_FILE, 'r', encoding='utf-8') as f:
            migration_sql = f.read()
        print(f'      Migration size: {len(migration_sql)} bytes')
        print()
    except FileNotFoundError:
        print(f'ERROR: Migration file not found: {MIGRATION_FILE}')
        sys.exit(1)

    # Connect to database
    print('[2/3] Connecting to VPS database...')
    print(f'      Host: {DB_CONFIG["host"]}')
    print(f'      Port: {DB_CONFIG["port"]}')
    print()

    try:
        conn = psycopg2.connect(**DB_CONFIG)
        conn.autocommit = False
        cursor = conn.cursor()
        print('      Connection established!')
        print()
    except Exception as e:
        print(f'ERROR: Could not connect to database')
        print(f'       {str(e)}')
        sys.exit(1)

    # Execute migration
    print('[3/3] Executing migration...')
    try:
        cursor.execute(migration_sql)
        conn.commit()
        print('      Migration executed successfully!')
        print()
    except Exception as e:
        conn.rollback()
        print(f'ERROR: Migration failed')
        print(f'       {str(e)}')
        cursor.close()
        conn.close()
        sys.exit(1)

    # Verify tables created
    print('Verifying tables...')
    try:
        cursor.execute("""
            SELECT table_name FROM information_schema.tables
            WHERE table_schema = 'public'
            AND table_name IN ('coupons', 'coupon_usages', 'payment_transactions', 'feature_manifests')
            ORDER BY table_name;
        """)
        tables = cursor.fetchall()
        print(f'  ✓ Found {len(tables)} new tables:')
        for table in tables:
            print(f'    - {table[0]}')
        print()
    except Exception as e:
        print(f'WARNING: Could not verify tables: {str(e)}')
        print()

    # Verify sample coupons
    print('Verifying sample coupons...')
    try:
        cursor.execute("SELECT code, discount_type, discount_value FROM coupons ORDER BY code;")
        coupons = cursor.fetchall()
        print(f'  ✓ Found {len(coupons)} sample coupons:')
        for code, dtype, value in coupons:
            print(f'    - {code} ({dtype}: {value})')
        print()
    except Exception as e:
        print(f'WARNING: Could not verify coupons: {str(e)}')
        print()

    # Clean up
    cursor.close()
    conn.close()

    print('=' * 60)
    print('SUCCESS: Migration 021 Applied Successfully!')
    print('=' * 60)
    print()
    print('Tables created:')
    print('  ✓ payment_transactions')
    print('  ✓ feature_manifests')
    print('  ✓ coupons')
    print('  ✓ coupon_usages')
    print('  ✓ referrals')
    print('  ✓ subscription_events')
    print()
    print('Functions created:')
    print('  ✓ check_entitlement()')
    print('  ✓ validate_coupon()')
    print('  ✓ generate_referral_code()')
    print()

if __name__ == '__main__':
    main()

File: apply-migration.mjs
import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';

const SUPABASE_URL = 'http://89.117.60.144:54321';
const SUPABASE_SERVICE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImV4cCI6MTk4MzgxMjk5Nn0.EGIM96RAZx35lJzdJsyH-qQwv8Hdp7fsn3W0YpN81IU';

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY, {
  auth: { persistSession: false }
});

console.log('=====================================');
console.log('Applying Migration 021');
console.log('=====================================\n');

console.log('[1/3] Reading migration file...');
const migrationSQL = readFileSync('packages/supabase/supabase/migrations/021_monetization_system.sql', 'utf8');
console.log(`      Migration size: ${migrationSQL.length} bytes\n`);

console.log('[2/3] Connecting to VPS database...');
console.log(`      URL: ${SUPABASE_URL}\n`);

console.log('[3/3] Executing migration...');

try {
  // Execute the migration SQL
  const { data, error } = await supabase.rpc('exec_sql', { sql_query: migrationSQL });

  if (error) {
    // Try alternative method - direct query
    console.log('      Trying alternative execution method...');

    const response = await fetch(`${SUPABASE_URL}/rest/v1/rpc/exec`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'apikey': SUPABASE_SERVICE_KEY,
        'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
        'Prefer': 'return=representation'
      },
      body: JSON.stringify({ query: migrationSQL })
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    console.log('      Alternative method succeeded!\n');
  } else {
    console.log('      Migration executed successfully!\n');
  }

  console.log('=====================================');
  console.log('SUCCESS: Migration 021 Applied!');
  console.log('=====================================\n');

  console.log('Tables created:');
  console.log('  ✓ payment_transactions');
  console.log('  ✓ feature_manifests');
  console.log('  ✓ coupons');
  console.log('  ✓ coupon_usages');
  console.log('  ✓ referrals');
  console.log('  ✓ subscription_events\n');

  console.log('Functions created:');
  console.log('  ✓ check_entitlement()');
  console.log('  ✓ validate_coupon()');
  console.log('  ✓ generate_referral_code()\n');

  console.log('Sample coupons:');
  console.log('  ✓ WELCOME20 (20% off)');
  console.log('  ✓ ANNUAL50 (50% off)');
  console.log('  ✓ FLAT100 (₹100 off)\n');

  // Verify tables exist
  const { data: tables, error: tableError } = await supabase
    .from('coupons')
    .select('code')
    .limit(3);

  if (!tableError && tables) {
    console.log(`Verification: Found ${tables.length} sample coupons`);
    tables.forEach(t => console.log(`  - ${t.code}`));
  }

  process.exit(0);

} catch (err) {
  console.error('\n=====================================');
  console.error('ERROR: Migration Failed');
  console.error('=====================================');
  console.error(err.message);
  console.error('\nFalling back to SQL file execution method...\n');

  // The SQL file can be manually applied via Supabase Studio
  console.log('Please apply migration manually via:');
  console.log(`1. Open: ${SUPABASE_URL.replace('54321', '3000')}`);
  console.log('2. Go to SQL Editor');
  console.log('3. Copy contents of: packages/supabase/supabase/migrations/021_monetization_system.sql');
  console.log('4. Paste and execute');

  process.exit(1);
}
